SRIOV Provider use bare sriov components instead of sriov-network-operator

ormergi opened the pull request:
Currently SRIOV provider we use sriov-network-operator to deploy sriov-cni and sriov-device-plugin
and configure PF's and VF's.
sriov-network-operator was meant to run on full flagged Openshift with bare-metal nodes that has
SRIOV hardware, and not on containerized Kubernetes (KIND) setup.

We need to add support to SRIOV provider to enable multiple SRIOV capable worker nodes setup
for Kubevirt SRIOV Live Migration feature.
At the moment we use version 4.4 where sriov-device-plugin is not compatible running on
containerized environments, causing VM's creation to fail due to VF allocation from other nodes.

This PR simplify the sriov provider by removing sriov-network-operator,
Instead it will configure PFs and VFs statically and deploy multus sriov-cni, sriov-device-plugin.

Supporting multiple SRIOV capable worker nodes will be introduces in a following PR.

ormergi commented:
/cc @EdDev

oshoval commented on cluster-up/cluster/kind-k8s-sriov-1.17.0/manifests/sriovdp-config.yaml:
```diff
@@ -0,0 +1,17 @@
+        "resourceName": "sriov_net",
+        "selectors": {
+          "drivers":["vfio-pci"],
+          "pfNames": ["enp4s0f0","enp4s0f1"]
```
this won't work on all machines, does it ?
locally i had to change it.

ormergi replied:
It is patched by config_sriov.sh later on, I added those interfaces names so one could have a sense of how it should look like.
I picked those names from what I saw on CI nodes

oshoval replied:
I suggest then to have a place holder instead them both, and envsubst it, wdyt?
it will be more self explanatory and maintainable

ormergi replied:
I am ok with using `envsubst` for now

oshoval commented:
Trying to run it with kubevirt locally, in order to check the multi sriov,
started by running just one


Removed this code from the test, since the resource is changed
```
		// Check if the hardware supports SRIOV
		// if err := validateSRIOVSetup(virtClient, sriovResourceName, 1); err != nil {
		// 	Skip("Sriov is not enabled in this environment. Skip these tests using - export FUNC_TEST_ARGS='--ginkgo.skip=SRIOV'")
		// }
```

I get this error in the virt-launcher
```
  Warning  FailedCreate  14s (x4 over 32s)  virtualmachine-controller  (combined from similar events): Error creating pod: Pod "virt-launcher-testvmi-8m9b9-5g467" is invalid: [spec.containers[0].resources.limits[sriov_net]: Invalid value: "sriov_net": must be a standard resource type or fully qualified, spec.containers[0].resources.limits[sriov_net]: Invalid value: "sriov_net": must be a standard resource for containers, spec.containers[0].resources.requests[sriov_net]: Invalid value: "sriov_net": must be a standard resource type or fully qualified, spec.containers[0].resources.requests[sriov_net]: Invalid value: "sriov_net": must be a standard resource for containers]
```

If you have please the manual patches that need to be done manually please let me know,
so i can test multi sriov locally and on CI with this PR.
Meanwhile i will try to debug if needed.

oshoval commented on cluster-up/cluster/kind-k8s-sriov-1.17.0/manifests/sriov-cni-daemonset.yaml:
```diff
@@ -0,0 +1,47 @@
+        effect: NoSchedule
+      containers:
+      - name: kube-sriov-cni
+        image: nfvpe/sriov-cni
```
This one is breathing, while the others are not
for example nfvpe/sriov-device-plugin:v3.3

worth to either let them both breath or freeze them both (i guess the 2nd is much better).
also best to put it on quay.io, else we hit rate limit, if no proxy is used (locally)
proxy with kind locally is not comfortable imo

ormergi replied:
I pushed a change now both images names and tags controlled from `Kustomization` manifest.

oshoval commented on cluster-up/cluster/kind-k8s-sriov-1.17.0/sriov_components.sh:
```diff
@@ -0,0 +1,184 @@
+  return 0
+}
+
+function sriov_components::deploy_sriov_components() {
```
yep it will also change that fact that there is _deploy_sriov_components and deploy_sriov_components

ormergi replied:
Done

oshoval commented on cluster-up/cluster/kind-k8s-sriov-1.17.0/sriov_components.sh:
```diff
@@ -0,0 +1,184 @@
+  return 0
+}
+
+function sriov_components::get_resource_name() {
```
in case this function can be refactored please it would be awesome,
lots of sed is little confusing

ormergi replied:
I changed how sriovdp config is patched, it uses `envsubst` now.

ormergi commented:
> Trying to run it with kubevirt locally, in order to check the multi sriov,
> started by running just one cluster
> 
> Removed this code from the test, since the resource is changed
> 
> ```
> 		// Check if the hardware supports SRIOV
> 		// if err := validateSRIOVSetup(virtClient, sriovResourceName, 1); err != nil {
> 		// 	Skip("Sriov is not enabled in this environment. Skip these tests using - export FUNC_TEST_ARGS='--ginkgo.skip=SRIOV'")
> 		// }
> ```
> 
> I get this error in the virt-launcher (and VM is Pending forever)
> 
> ```
>   Warning  FailedCreate  14s (x4 over 32s)  virtualmachine-controller  (combined from similar events): Error creating pod: Pod "virt-launcher-testvmi-8m9b9-5g467" is invalid: [spec.containers[0].resources.limits[sriov_net]: Invalid value: "sriov_net": must be a standard resource type or fully qualified, spec.containers[0].resources.limits[sriov_net]: Invalid value: "sriov_net": must be a standard resource for containers, spec.containers[0].resources.requests[sriov_net]: Invalid value: "sriov_net": must be a standard resource type or fully qualified, spec.containers[0].resources.requests[sriov_net]: Invalid value: "sriov_net": must be a standard resource for containers]
> ```
> 
> If you have please the manual patches that need to be done manually please let me know,
> so i can test multi sriov locally and on CI with this PR.
> Meanwhile i will try to debug if needed.
> 
> You used to call the resource kubevirt.io/sriov_net isn't it ?
> now its just sriov_net i think (not sure)
> EDIT trying to change to `kubevirt.io/sriov_net`
> 
> I will also first try it without any of my changes, but they shouldn't affect this.
> EDIT2
> with `kubevirt.io/sriov_net` running one at a time works.

As you figured out on kubevirt tests you need to change `sriovResourceName` from `openfhost.io/...` to `kubevirt.io/...`

oshoval commented:
Multi sriov passed on CI using this PR + some other needed PRs (some that are not merged yet)

Running 3 jobs at the time
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/851/rehearsal-pull-kubevirt-e2e-kind-1.17-sriov/1369929074130030592
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/853/rehearsal-pull-kubevirt-e2e-kind-1.17-sriov/1369929074138419200
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/854/rehearsal-pull-kubevirt-e2e-kind-1.17-sriov/1369929074142613504

ormergi commented:
Changes:
- Addressed review comments:
- Now its possible to control:
  - `sriovdp` (sriov-device-plugin) and `sriov-cni `nodeSelector label
  - `sriovdp` (VF's pool) resource name and prefix.
  -  The driver `sriovdp` will discovered VF's to be configured with.
From the script instead of relaying on static config.
These parameters passed to `sriov_components:deploy`.
Driver parameter is passed to `configure_vfs.sh` script 
along with the driver module name.
- sriov-components related files moved to new directory  "sriov-components"
- Remove sourcing kind/common.sh from config_sriov.sh

EdDev commented on cluster-up/cluster/kind-k8s-sriov-1.17.0/provider.sh:
```diff
@@ -25,7 +20,7 @@ function up() {
     kind_up
 
-    # remove the rancher.io kind default storageClass
+     remove the rancher.io kind default storageClass
```
The comment was removed?

ormergi replied:
No, just a mistake
I will bring it back

EdDev commented on cluster-up/cluster/kind-k8s-sriov-1.17.0/provider.sh:
```diff
@@ -11,11 +11,6 @@ function set_kind_params() {
 }
 
 function up() {
-    if [[ "$KUBEVIRT_NUM_NODES" -ne 2 ]]; then
```
How is this removal related to this commit?

ormergi replied:
Its not related I will bring it back sorry

EdDev reviewed (commented):
Commit "Present sriov_components.sh script":

Typo: `wtih`

ormergi commented:
/test check-up-kind-1.17-sriov

ormergi commented:
/test check-up-kind-1.17-sriov

qinqon commented on cluster-up/cluster/kind-k8s-sriov-1.17.0/sriov_operator.sh:
```diff
@@ -0,0 +1,212 @@
+  return 0
+}
+
+function sriov_operator::deploy_sriov_operator {
```
this is not used

qinqon commented on cluster-up/cluster/kind-k8s-sriov-1.17.0/sriov_operator.sh:
```diff
@@ -0,0 +1,212 @@
+  return 0
+}
+
+function sriov_operator::apply_sriov_node_policy {
```
This is not used

qinqon reviewed (changes requested):
First pass, some unused functions, /hold let's wait for https://github.com/kubevirt/kubevirt/pull/5286

ormergi commented:
> First pass, some unused functions, /hold let's wait for [kubevirt/kubevirt#5286](https://github.com/kubevirt/kubevirt/pull/5286)

Hey I didnt remove anything that related to sriov-network-operator until we happy with sriov-cni and sriovdp alone
so we could revert in case of regration

ormergi commented:
SRIOV lane is grean with this PR on kubevirt 
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/5286/pull-kubevirt-e2e-kind-1.17-sriov/1373999101787508736

oshoval commented:
Please verify this PR on all the 3 CI nodes.

ormergi commented:
> Please verify this PR on all the 3 CI nodes.

Is there a way to run check-up-kind-1.17-sriov on specific node?

ormergi commented:
https://prow.apps.ovirt.org/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/563/check-up-kind-1.17-sriov/1372224155977519104
CI node srv20

qinqon commented:
> Please verify this PR on all the 3 CI nodes.
> 
> 

Why do we need that? Aren't the SRIOV nodes homogenous?

EdDev commented:
> > Please verify this PR on all the 3 CI nodes.
> 
> Why do we need that? Aren't the SRIOV nodes homogenous?

An explicit reason should be given to do such a HW dependent check, I agree.

The nodes are note homogeneous in terms of HW specs, e.g. the number of NIC/s and their name may be different.
But if I am not mistaken, this change does not depend on a specific NIC name.

qinqon commented:
> > > Please verify this PR on all the 3 CI nodes.
> > 
> > 
> > Why do we need that? Aren't the SRIOV nodes homogenous?
> 
> An explicit reason should be given to do such a HW dependent check, I agree.
> 
> The nodes are not homogeneous in terms of HW specs, e.g. the number of NIC/s and their name may be different.
> But if I am not mistaken, this change does not depend on a specific NIC name.

Then they are "homogeneous enough" to approve this, what do you think ?

EdDev commented:
> Then they are "homogeneous enough" to approve this, what do you think ?

We can wait for tomorrow to hear the reasoning in more details.

oshoval commented:
I saw one time that srv24 didn't cluster up with this PR, don't have it in hand.

Each machine of CI has different card.
I think such change as this PR presents, worth running at least 2 times on each machine of CI to see it works.
Up to you.

ormergi commented:
> I saw one time that srv24 didn't cluster up with this PR, don't have it in hand.

It failed because the job was terminated, it didnt got into the interesting part
https://prow.apps.ovirt.org/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/563/check-up-kind-1.17-sriov/1369585181609955328

> Each machine of CI has different card.
> I think such change as this PR presents, worth running at least 2 times on each machine of CI to see it works.
> Up to you.

How to invoke a job on specific job? I am almost sure you had to do the same for multi cluster PR's.
I am asking because I am trying to avoid blindly retesting the job

oshoval commented:
Just run it several times please for the time being imo.

ormergi commented:
/test check-up-kind-1.17-sriov

ormergi commented:
Passed on all CI nodes
srv17
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/5286/pull-kubevirt-e2e-kind-1.17-sriov/1373999101787508736
srv20
https://prow.apps.ovirt.org/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/563/check-up-kind-1.17-sriov/1372224155977519104
srv24
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/563/check-up-kind-1.17-sriov/1374741619265245184

qinqon commented:
/lgtm
/approve

qinqon commented:
/retest

Not related

```
deployment.apps/virt-operator created
+ ./cluster-up/kubectl.sh apply -f https://storage.googleapis.com/kubevirt-prow/devel/nightly/release/kubevirt/kubevirt/20210325/kubevirt-cr.yaml
error: unable to recognize "https://storage.googleapis.com/kubevirt-prow/devel/nightly/release/kubevirt/kubevirt/20210325/kubevirt-cr.yaml": no matches for kind "KubeVirt" in version "kubevirt.io/v1"
++ cleanup
` ``

ormergi commented:
> /lgtm
> /approve

Thanks you!

ormergi commented:
/retest

Not related

ormergi commented:
/retest
